[Feature-2925][server] Init TaskLogger in TaskExecuteProcessor (#2925)

yangyichao-mango opened the pull request:
## What is the purpose of the pull request

*Init task logger before init TaskExecuteThread and pass the `taskLogger` in to TaskExecuteThread init method as parameter, so we can record the user and work dir creating log to task instance log, and user can read this log in front-end WebUI.* #3014 
*Use common ShellExecutor in FileUtils.* #2925 

## Brief change log

  - *Add `taskLogger` parameter  to TaskExecuteThread.* 
  - *Add three log utils to LoggerUtils.*
  - *Init task logger before create initing TaskExecuteThread, and init task logger process to TaskExecuteProcessor.*
  - *Log all detail when create user and working dir in FileUtils.*

## Verify this pull request

This pull request is code cleanup without any test coverage.

yangyichao-mango commented:
Hi @gabrywu @dailidong , Can you help to review this PR? Thx a lot.

gabrywu commented:
@yangyichao-mango What's purpose of this PR? Why are there so many irrelevant files to change ?

yangyichao-mango commented:
> @yangyichao-mango What's purpose of this PR? Why are there so many irrelevant files to change ?

Thx a lot for your review @gabrywu .

This PR is essentially to solve the abnormal exit caused by the error (exceptions caused by permissions, etc.) when the worker starts to create the working directory

But after my change, I found that if the worker failed to create the working directory, the exception it threw would be caught, only recorded in the log of the worker node, not in the log of a single task instance. Although the task will fail at the end of execution, the exception log is very few and it is difficult to locate the exception (permission exception, etc.) directly. 

In my opinion, I understand that this error should fail fast, or at least recorded by logs. So in order to match the change of my exitval judgment work, I put the initialization change of tasklogger to  the beginning of `TaskExecuteProcessor`. And I understand that the task logger should record all the logs related to the task, not only the task runtime logs.

If I have any misunderstanding, please give me suggestions.

------------------------------

非常感谢你的review。

这个pr本质上是为了解决在worker执行task时创建工作目录时的错误（由于权限导致的问题等）导致的异常退出。

但是在改动完成后，我发现如果worker创作工作目录失败其抛出的异常会被catch住，异常信息只会在worker节点的日志中被记录下来，不会在单个task的日志中记录下来，这相当于我一开始的工作完全不会被外显。虽然最后task执行的时候会失败，但是异常日志非常少，很难直接定位到工作目录中的异常（权限异常）。

我理解这种错误应该fail-fast，或者至少有日志来记录，所以为了配合我这次针对exitval判断的改动，我就把taskLogger的初始化改动到了最前面，用来记录涉及到这个task的所有日志。并且我理解taskLogger本身就应该记录所有涉及到这个task的日志，不仅仅是task运行时的日志。

如果我有理解不正确的地方，敬请指出。

gabrywu commented:
> > @yangyichao-mango What's purpose of this PR? Why are there so many irrelevant files to change ?
> 
> Thx a lot for your review @gabrywu .
> 
> This PR is essentially to get the abnormal exit value caused by the error (exceptions caused by permissions, etc.) when the worker starts to create the working directory, to judge if the user info and work dir is created successfully.
> 
> But after my change, I found that if the worker failed to create the working directory, the exception it threw would be caught, only recorded in the log of the worker node, not in the log of a single task instance. Although the task will fail at the end of execution, the exception log is very few and it is difficult to locate the exception (permission exception, etc.) directly.
> 
> In my opinion, I understand that this error should fail fast, or at least recorded by logs. So in order to match the change of my exitval judgment work, I put the initialization change of tasklogger to the beginning of `TaskExecuteProcessor`. And I understand that the task logger should record all the logs related to the task, not only the task runtime logs.
> 
> If I have any misunderstanding, please give me suggestions.
> 
> 非常感谢你的review。
> 
> 这个pr本质上是为了获取在worker执行task时创建工作目录时的错误（由于权限导致的问题等）值，来判断工作目录，用户权限是否创建成功。
> 
> 但是在改动完成后，我发现如果worker创作工作目录失败其抛出的异常会被catch住，异常信息只会在worker节点的日志中被记录下来，不会在单个task的日志中记录下来，这相当于创建工作目录相关的所有错误信息完全不会被直接外显给用户。虽然最后task执行的时候会失败，但是异常日志非常少，很难直接定位到工作目录中的异常（权限异常等）。
> 
> 我理解这种错误应该fail-fast，或者至少有日志来记录，所以为了配合我这次针对exitval判断的改动，我就把taskLogger的初始化改动到了最前面，用来记录涉及到这个task的所有日志。并且我理解taskLogger的原本目的就是用来记录所有涉及到这个task的日志，不仅仅是task运行时的日志。
> 
> 如果我有理解不正确的地方，敬请指出。

You'd better give a suitable title to this issue, otherwise others may have a misunderstanding

yangyichao-mango commented:
> > > @yangyichao-mango What's purpose of this PR? Why are there so many irrelevant files to change ?
> > 
> > 
> > Thx a lot for your review @gabrywu .
> > This PR is essentially to get the abnormal exit value caused by the error (exceptions caused by permissions, etc.) when the worker starts to create the working directory, to judge if the user info and work dir is created successfully.
> > But after my change, I found that if the worker failed to create the working directory, the exception it threw would be caught, only recorded in the log of the worker node, not in the log of a single task instance. Although the task will fail at the end of execution, the exception log is very few and it is difficult to locate the exception (permission exception, etc.) directly.
> > In my opinion, I understand that this error should fail fast, or at least recorded by logs. So in order to match the change of my exitval judgment work, I put the initialization change of tasklogger to the beginning of `TaskExecuteProcessor`. And I understand that the task logger should record all the logs related to the task, not only the task runtime logs.
> > If I have any misunderstanding, please give me suggestions.
> > 非常感谢你的review。
> > 这个pr本质上是为了获取在worker执行task时创建工作目录时的错误（由于权限导致的问题等）值，来判断工作目录，用户权限是否创建成功。
> > 但是在改动完成后，我发现如果worker创作工作目录失败其抛出的异常会被catch住，异常信息只会在worker节点的日志中被记录下来，不会在单个task的日志中记录下来，这相当于创建工作目录相关的所有错误信息完全不会被直接外显给用户。虽然最后task执行的时候会失败，但是异常日志非常少，很难直接定位到工作目录中的异常（权限异常等）。
> > 我理解这种错误应该fail-fast，或者至少有日志来记录，所以为了配合我这次针对exitval判断的改动，我就把taskLogger的初始化改动到了最前面，用来记录涉及到这个task的所有日志。并且我理解taskLogger的原本目的就是用来记录所有涉及到这个task的日志，不仅仅是task运行时的日志。
> > 如果我有理解不正确的地方，敬请指出。
> 
> You'd better give a suitable title to this issue, otherwise others may have a misunderstanding. If you want to fix different issue, different PRs should be created

Thx a lot for your review @gabrywu .
I've change the description of this issue and PR and connect the [FEATURE] #3014 and [BUG] #2925.

dailidong commented on dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/FileUtils.java:
```diff
@@ -138,24 +153,39 @@ public static String getResourceViewSuffixs() {
+                    logger.error(infoLog);
+                }
+            }
+        } catch (Throwable ignored) {
         }
```
why swallow exception？

yangyichao-mango replied:
> why swallow exception？

Thx a lot for your review @dailidong . I've removed the unused exception try catch clause.

dailidong commented:
What is the function of this ThreadLocal? Is it necessary? I think we changed our log format and implemented it with loggerutils

yangyichao-mango commented:
> What is the function of this ThreadLocal? Is it necessary? I think we changed our log format and implemented it with loggerutils

Thx a lot for your review @dailidong .
In my opinion, it's better to set the `tasklogger` threadlocal value to log the task detail but not to pass the `tasklogger` through the function's input params in the utils class of `FileUtils` and `OSUtils`.

yangyichao-mango commented:
> What is the function of this ThreadLocal? Is it necessary? I think we changed our log format and implemented it with loggerutils

Thx a lot for your review and work of merging dev branch to current branch @dailidong .
In my opinion, although my branch is out-of-date with the base branch, maybe it's better to rebase the current branch. If it's needed, I will rebase my branch after your approval this PR and reply me.

dailidong commented:
I think no need to rebase if no confilct files

yangyichao-mango commented:
> I think no need to rebase if no confilct files

Thx a lot for your review. @dailidong

chuyang-zl commented:
Hello, there is still a problem under the modified logic:
If you copy a resource file and there is no exception (the file is in the resource list, but not in HDFS), you will still not see the relevant error message in the task log of the Web UI, making it difficult to troubleshoot the problem

@yangyichao-mango 
---

您好，在现在修改后的逻辑下，还存在一个问题是：
如果在copy资源文件时发生文件不存在异常（文件在资源列表中，但是在hdfs上不存在的特殊情况下），那么在web UI的task日志中还是看不到相关错误信息，不便于排查问题
![图片](https://user-images.githubusercontent.com/56599784/104543199-007d6880-5660-11eb-89e8-55d689e2c113.png)
